src/app/clusters/tls-client-management-server/tls-client-management-server.cpp

```diff
@@ -0,0 +1,244 @@
+    AttributeAccessInterface(MakeOptional(endpointId), TlsClientManagement::Id),
+    CommandHandlerInterface(MakeOptional(endpointId), TlsClientManagement::Id), mDelegate(delegate), mCertificateTable(certificateTable), mMaxProvisioned(0)
+{
+    mDelegate.SetTlsClientManagementServer(this);
+}
+
+TlsClientManagementServer::~TlsClientManagementServer()
+{
+    // null out the ref to us on the delegate
+    mDelegate.SetTlsClientManagementServer(nullptr);
+
+    // unregister
+    CommandHandlerInterfaceRegistry::Instance().UnregisterCommandHandler(this);
+    AttributeAccessInterfaceRegistry::Instance().Unregister(this);
+}
+
+CHIP_ERROR TlsClientManagementServer::Init()
+{
+    mCertificateTable.Init(Server::GetInstance().GetPersistentStorage());
+
+    LoadPersistentAttributes();
+
+    VerifyOrReturnError(AttributeAccessInterfaceRegistry::Instance().Register(this), CHIP_ERROR_INTERNAL);
+    ReturnErrorOnFailure(CommandHandlerInterfaceRegistry::Instance().RegisterCommandHandler(this));
+
+    return CHIP_NO_ERROR;
+}
+
+
+CHIP_ERROR TlsClientManagementServer::Finish()
+{
+    mCertificateTable.Finish();
+    return CHIP_NO_ERROR;
+}
+
+void TlsClientManagementServer::LoadPersistentAttributes()
+{
+    // Load MaxProvisioned
+    uint8_t storedMaxProvisioned = 5;
+    CHIP_ERROR err               = GetSafeAttributePersistenceProvider()->ReadScalarValue(
+        ConcreteAttributePath(GetEndpointId(), TlsClientManagement::Id, MaxProvisioned::Id), storedMaxProvisioned);
+    if (err != CHIP_NO_ERROR)
+    {
+        mMaxProvisioned = storedMaxProvisioned;
+    }
+    else
+    {
+        // otherwise defaults
+        ChipLogDetail(
+            Zcl, "TlsClientManagement: Unable to load the MaxProvisioned attribute from the KVS. Defaulting to %u",
+            mMaxProvisioned);
+    }
+}
+
+// AttributeAccessInterface
+CHIP_ERROR TlsClientManagementServer::Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder)
+{
+    VerifyOrDie(aPath.mClusterId == TlsClientManagement::Id);
+
+    switch (aPath.mAttributeId)
+    {
+    case MaxProvisioned::Id:
+        return aEncoder.Encode(mMaxProvisioned);
+    case ProvisionedEndpoints::Id:
+        TlsClientManagementServer * server = this;
+        CHIP_ERROR err =
+            aEncoder.EncodeList([server](const auto & encoder) -> CHIP_ERROR { return server->EncodeProvisionedEndpoints(encoder); });
+        return err;
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+uint8_t TlsClientManagementServer::GetMaxProvisioned() const
+{
+    return mMaxProvisioned;
+}
+
+// helper method to get the TlsClientManagement Sounds one by one and encode into a list
+CHIP_ERROR
+TlsClientManagementServer::EncodeProvisionedEndpoints(const AttributeValueEncoder::ListEncodeHelper & encoder)
+{
+    for (uint8_t i = 0; true; i++)
+    {
+        TLSEndpointStruct::Type endpoint;
+
+        auto err = mDelegate.GetProvisionedEndpointByIndex(i, endpoint);
```

Comment:
What does the lifetime management for the Hostname field of the struct look like?

That is, how does our delegate know when it's safe to free that memory?

Reply:
Fixed